synapse/handlers/federation.py

```diff
@@ -1343,32 +1343,50 @@ async def exchange_third_party_invite(
-                room_version_obj, event_dict, event, context
-            )
+            # Try 2 times, the first one could fail with PartialStateConflictError
+            # in send_membership_event, cf comment in except block.
+            for _ in range(2):
+                try:
+                    (
+                        event,
+                        context,
+                    ) = await self.event_creation_handler.create_new_client_event(
+                        builder=builder
+                    )
 
-            EventValidator().validate_new(event, self.config)
+                    event, context = await self.add_display_name_to_third_party_invite(
+                        room_version_obj, event_dict, event, context
+                    )
 
-            # We need to tell the transaction queue to send this out, even
-            # though the sender isn't a local user.
-            event.internal_metadata.send_on_behalf_of = self.hs.hostname
+                    EventValidator().validate_new(event, self.config)
 
-            try:
-                validate_event_for_room_version(event)
-                await self._event_auth_handler.check_auth_rules_from_context(event)
-            except AuthError as e:
-                logger.warning("Denying new third party invite %r because %s", event, e)
-                raise e
+                    # We need to tell the transaction queue to send this out, even
+                    # though the sender isn't a local user.
+                    event.internal_metadata.send_on_behalf_of = self.hs.hostname
 
-            await self._check_signature(event, context)
+                    try:
+                        validate_event_for_room_version(event)
+                        await self._event_auth_handler.check_auth_rules_from_context(
+                            event
+                        )
+                    except AuthError as e:
+                        logger.warning(
+                            "Denying new third party invite %r because %s", event, e
+                        )
+                        raise e
 
-            # We retrieve the room member handler here as to not cause a cyclic dependency
-            member_handler = self.hs.get_room_member_handler()
-            await member_handler.send_membership_event(None, event, context)
+                    await self._check_signature(event, context)
+
+                    # We retrieve the room member handler here as to not cause a cyclic dependency
+                    member_handler = self.hs.get_room_member_handler()
+                    await member_handler.send_membership_event(None, event, context)
+
+                    break
+                except PartialStateConflictError:
+                    # Persisting couldn't happen because the room got un-partial stated
+                    # in the meantime and context needs to be recomputed, so let's do so.
+                    pass
```

Comment:
I'm not 100% sure that we can't get a `PartialStateConflictError` the second time, since we might be subject to unlikely races where we leave the room, then rejoin. If at all possible, could we reraise the exception on the 2nd iteration?

(applies to all places)

Reply:
Oh yes I didn't think about that. My first version was using a while `loop` instead of a `for range(2)`, perhaps I should go back to that ?

I moved to `for` out of precaution to avoid infinite loop, but it really shouldn't happen, in the worst case we should get another exception and this one would be raised and would avoid infinite loop.

Otherwise I can do what you said but it still seems to me like we are throwing an implementation detail at the client, they shouldn't have to deal with that if it's a normal corner case I believe.

Reply:
I've switch to a while loop, this also avoids mypy complaining on some missing return for `_upgrade_room`.